mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/logger/MavenExtensionLoggerFactory.java

```diff
@@ -0,0 +1,71 @@
+package org.arquillian.smart.testing.mvn.ext.logger;
+
+import org.arquillian.smart.testing.configuration.Configuration;
+import org.arquillian.smart.testing.logger.DefaultLoggerFactory;
+import org.arquillian.smart.testing.logger.Logger;
+import org.arquillian.smart.testing.logger.LoggerFactory;
+
+public class MavenExtensionLoggerFactory implements LoggerFactory {
+
+    private final org.codehaus.plexus.logging.Logger logger;
+    private Configuration configuration;
+
+    public MavenExtensionLoggerFactory(org.codehaus.plexus.logging.Logger logger, Configuration configuration) {
+        this.logger = logger;
+        this.configuration = configuration;
+    }
+
+    public MavenExtensionLoggerFactory(org.codehaus.plexus.logging.Logger logger) {
+        this.logger = logger;
+    }
+
+    @Override
+    public Logger getLogger() {
+        return new MavenExtensionLogger(logger, configuration);
+    }
+
+    private class MavenExtensionLogger implements Logger {
+
+        private static final String SMART_TESTING_EXTENSION_PREFIX = "Smart Testing Extension - ";
+        private final org.codehaus.plexus.logging.Logger logger;
+        private Configuration configuration;
+
+        MavenExtensionLogger(org.codehaus.plexus.logging.Logger logger, Configuration configuration) {
+            this.logger = logger;
+            this.configuration = configuration;
+        }
+
+        @Override
+        public void info(String msg, Object... args) {
+            logger.info(getFormattedMsg(msg, args));
+        }
+
+        @Override
+        public void warn(String msg, Object... args) {
+            logger.warn(getFormattedMsg(msg, args));
+        }
+
+        @Override
+        public void debug(String msg, Object... args) {
+            logger.debug(getFormattedMsg(msg, args));
+            if (configuration != null && configuration.isDebug() && !logger.isDebugEnabled()) {
+                new DefaultLoggerFactory(true).getLogger().debug(msg, args);
+            }
+        }
+
+        @Override
+        public void error(String msg, Object... args) {
+            logger.error(getFormattedMsg(msg, args));
+        }
+
+        @Override
+        public boolean isDebug() {
+            return logger.isDebugEnabled() || (configuration != null && configuration.isDebug())
+                   || Boolean.valueOf(System.getProperty(Configuration.SMART_TESTING_DEBUG));
```

Comment:
We don't need to check this here as configuration is already overwriting it if `smart.testing.debug` is set using system property.

Reply:
It is necessary to check it here. We set the `MavenExtensionLoggerFactory` before we create the configuration object: https://github.com/hemanik/smart-testing/blob/be0c6c9deba19d50f075fdc0f40a658d72c63372/mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/SmartTestingMavenConfigurer.java#L52 and we don't pass the configuration as a parameter, which means that it is null. So we can only check the system property.
Then, when the configuration instance is retrieved, we can reset the factory with a new instance that contains also the `Configuration`: https://github.com/hemanik/smart-testing/blob/be0c6c9deba19d50f075fdc0f40a658d72c63372/mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/SmartTestingMavenConfigurer.java#L55

Reply:
But in this case if user is not setting system property & he is defining `debug: true` in config file. then in this case we should enable debug options, but we are not enabling it because configuration object is not created before setting `MavenExtensionLoggerFactory`

Reply:
Yeah, exactly. This is a chicken-egg problem and there is no better solution for it.
BTW, currently, there is no debug log in the `Configuration` class - only info & warn, so you don't have to care about it - we don't miss any information ;-)

Reply:
This is fine for now. But we should add comment about this why we did this way, so any new developer can understand it easily